### PR TITLE
Use the right symbol for constant string placeholder

### PR DESCRIPTION
**Pull request by liqunl**
Constant string can be used as a placeholder and be patched to other
object. Set the flag on the symbol according to the actual object type.

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by DanHeidinga** [comment]
@liqunl my understanding was that other values besides MHs can be stored in these "string" constants as the Unsafe.defineAnonymousClass allows any object to be stored there.  The bytecode pattern will be something like:
```
ldc  #42 // Constant_String but patched to some other object
instanceof <real type>
....
```
We want to be careful to not rely on the resolved value only being a string or a MH.

@fengxue-IS / @babsingh to confirm

**Comment by fengxue-IS** [comment]
for Java 8-14 where the InvokerBytecodeGenerator uses cp patching to reference dynamic object, the constant string placeholder can contain non-MH objects.
for Java 15+, the patch objects are stored into hiddenclass's classData field, so doesn't require checkcast.
@babsingh to confirm about the hiddenclass classData access

**Comment by babsingh** [comment]
> to confirm about the hiddenclass classData access

In JDK15+, `InvokerBytecodeGenerator` uses the below-mentioned fields of `ClassData` to generate the appropriate bytecodes with the required `checkcast` instructions. In the below bytecodes, the static init block loads the `value` from `classData` into the field (`_DATA_LambdaForm_0`) with the correct `checkcast`. **So, a `checkcast` won't be needed during invocation in JDK15+.**

An **instance of `ClassData`** contains:
```
        // Used to generate the field to access the object; e.g. _DATA_LambdaForm_0.
        final String name;

        // Describes if the object is a Class, MethodHandle, LambdaForm or Object.
        // Used to add a checkcast before storing the object value into the generated field.
        final String desc;

        // Stored object value.
        final Object value;
```

**Sample generated bytecodes** to access `classData` in JDK16:
```
  static {};
    descriptor: ()V
    flags: (0x0008) ACC_STATIC
    Code:
      stack=2, locals=1, args_size=0
         0: ldc           #2                  // class java/lang/invoke/LambdaForm$BMH003
         2: invokestatic  #32                 // Method java/lang/invoke/MethodHandleNatives.classData:(Ljava/lang/Class;)Ljava/lang/Object;
         5: checkcast     #34                 // class java/util/List
         8: astore_0
         9: aload_0
        10: iconst_0
        11: invokeinterface #38,  2           // InterfaceMethod java/util/List.get:(I)Ljava/lang/Object;
        16: checkcast     #40                 // class java/lang/invoke/LambdaForm
        19: putstatic     #42                 // Field _DATA_LambdaForm_0:Ljava/lang/invoke/LambdaForm;
        22: return
```

**Comment by liqunl** [comment]
@DanHeidinga @fengxue-IS @babsingh Thanks for pointing that out. Will change the code to load any objects.

**Comment by liqunl** [comment]
Depends on https://github.com/eclipse/omr/pull/5664

**Comment by liqunl** [comment]
@andrewcraik @DanHeidinga May I ask for your review?
@dmitry-ten Could you review the JITServer change?

File: runtime/compiler/exceptions/AOTFailure.hpp
Comment (andrewcraik) [inline runtime/compiler/exceptions/AOTFailure.hpp:98]: Question - should this be recoverable? We can recover only with a non-AOT compile so wanted ot make sure this is the right semantics

**Review by andrewcraik — APPROVED** [review]
The non-JITServer parts look good to me

File: runtime/compiler/net/MessageTypes.hpp
Comment (dmitry-ten) [inline runtime/compiler/net/MessageTypes.hpp:138]: Could you please update message numbers that come after this message?

File: runtime/compiler/net/MessageTypes.hpp
Comment (dmitry-ten) [inline runtime/compiler/net/MessageTypes.hpp:411]: Same here, please update message numbers.

**Review by dmitry-ten — COMMENTED** [review]
Overall LGTM,
please increment `MINOR_NUMBER` in `JITServer::CommunicationStream` and update message numbers.

**Comment by dmitry-ten** [comment]
This commit results in 2 additional messages when a string is unresolved but if the increase is significant we'll deal with it later.

File: runtime/compiler/exceptions/AOTFailure.hpp
Comment (liqunl) [reply runtime/compiler/exceptions/AOTFailure.hpp:98]: Yes.  Added code to disable AOT in recompilation.

File: runtime/compiler/net/MessageTypes.hpp
Comment (liqunl) [reply runtime/compiler/net/MessageTypes.hpp:138]: Updated

File: runtime/compiler/net/MessageTypes.hpp
Comment (liqunl) [reply runtime/compiler/net/MessageTypes.hpp:411]: Updated

**Comment by liqunl** [comment]
> This commit results in 2 additional messages when a string is unresolved but if the increase is significant we'll deal with it later.

It's 2 additional messages when the cp entry is resolved. I can fix it by checking if the symref is first created, only do the check if symref is first created.

**Comment by dmitry-ten** [comment]
> It's 2 additional messages when the cp entry is resolved. I can fix it by checking if the symref is first created, only do the check if symref is first created.
If that's something that would potentially reduce the number of messages then sure, it'd be nice.

**Comment by liqunl** [comment]
Looking at a failure in my pbuild after updating the commits, WIP the PR now.

**Comment by liqunl** [comment]
Fixed the pbuild failure. @dmitry-ten @andrewcraik Could you take a look again?

**Comment by dmitry-ten** [comment]
Please increment the minor number here:
https://github.com/eclipse/openj9/blob/547bb8e04985d8043443b977ef979eb183139da6/runtime/compiler/net/CommunicationStream.hpp#L112
Otherwise, LGTM

**Review by andrewcraik — APPROVED** [review]

**Comment by liqunl** [comment]
@dmitry-ten Fixed the minor number

**Comment by dmitry-ten** [comment]
LGTM

**Comment by liqunl** [comment]
@DanHeidinga I have made changes to support any type of patched objects, could you review again?

**Comment by DanHeidinga** [comment]
> @DanHeidinga I have made changes to support any type of patched objects, could you review again?

Design wise, this seems right to me.  I'll defer to @andrewcraik's already provided review.

**Comment by andrewcraik** [comment]
LGTM

**Comment by liqunl** [comment]
@DanHeidinga Andrew has approved this PR, may I ask you to help launching the PR build and merge if it passes?

**Comment by DanHeidinga** [comment]
Jenkins test sanity all jdk11

**Comment by DanHeidinga** [comment]
Builds have passed and given @andrewcraik's approval above, merging

**Pull request merged**
